### ISSUE TITLE
add noise to weights or gradients during training

### ISSUE THREAD

**Issue by hma02** [tensorflow/tfjs-core#313]
Hello,

I'm trying the predicting complementary color demo and have noticed that it happens quite often that the training get stuck in some minima resulting in inferencing zeros in (r,g,b) channels, which slows down or even stops training progress.

Now I wonder if it helps to add certain scale of random noise to weights/gradients during training. 

Any hints on how can I get started with this? I looked through the SGD optimizer and found [this line](https://github.com/PAIR-code/deeplearnjs/blob/master/src/graph/optimizers/sgd_optimizer.ts#L39):

```
const variable = math.scaledArrayAdd(this.c, gradient, this.one, oldVariable);
```

Should I proceed to add a random number here? Something like:

```
const variable = math.scaledArrayAdd(this.one, NDArray.randNormal(weightsShape, mean, stdev), this.one, variable)
```

Also I wonder if there is a Tensor type random number generator? Thanks

**Comment by nsthorat** [comment]
There is no Tensor type random number unfortunately. If you'd like to add it, I can give you pointers. Let me know if that interests you.

**Comment by DiThi** [comment]
I'm interested in this as well.

IIRC there's a similar feature in TF called dropout for a similar purpose, which would also be interesting to have in dl.js and also relies in random numbers.

**Comment by hma02** [comment]
@nsthorat ,

I was able to hacking the code and create a temporary `noiseInjectOptimizer` to be used in `math.scope` and found that injecting noise actually helps.  But simply using a larger learning rate can reduce the chances of bad initialization. 

![log](https://user-images.githubusercontent.com/12686013/33131538-d48a254a-cf64-11e7-80ff-2a1708838ec8.PNG)
![chart](https://user-images.githubusercontent.com/12686013/33131540-d7893c68-cf64-11e7-9be7-0450c0556e70.png)

**Comment by nsthorat** [comment]
Dropout is now added in TensorFlow.js layers.
